### PR TITLE
Escarrá HW02

### PR DESCRIPTION
**Pull request by Alain-David-001**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -7,148 +7,156 @@
...
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex a = zipWith (^) a [0 ..]
 
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference [] = error "List cannot be empty"
+productOfDifference [a] = 1
+productOfDifference (x : xs) = productOfDifference xs * (x - head xs)
```
Comment (alex28sh) [inline Main.hs:22]: What about using `zipWith`?

File: Main.hs [hunk 2]
```diff
@@ -7,148 +7,156 @@
...
 
-isSorted :: [Int] -> Bool 
-isSorted xs = undefined
+isSorted :: [Int] -> Bool
+isSorted [] = True
+isSorted [a] = True
+isSorted (x:xs) = x <= head xs && isSorted xs
```
Comment (alex28sh) [inline Main.hs:27]: What about higher-order functions?

File: Main.hs [hunk 3]
```diff
@@ -7,148 +7,156 @@
...
-countElement = undefined
+countElement :: Int -> [Int] -> Int
+countElement n xs= length (filter (== n) xs)
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct a [] = 0
+dotProduct [] a = 0
+dotProduct (a:as) (b:bs) = a*b + dotProduct as bs
```
Comment (alex28sh) [inline Main.hs:35]: again

File: Main.hs [hunk 4]
```diff
@@ -7,148 +7,156 @@
...
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll fs x = map (\func -> func x) fs
 
-interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave :: [a] -> [a] -> [a]
+interleave a [] = []
+interleave [] a = []
+interleave as bs = concat (zipWith (\ a b -> [a,b]) as bs)
```
Comment (alex28sh) [inline Main.hs:43]: you don't need first 2 pattern-matchings

**Review by alex28sh — CHANGES_REQUESTED** [review]
4/8

File: Main.hs [hunk 5]
```diff
@@ -7,148 +7,150 @@
...
+multByIndex = zipWith (*) [0 ..]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex a = zipWith (^) a [0 ..]
 
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference [] = error "List cannot be empty"
+productOfDifference [a] = error "List cannot be empty"
```
Comment (kajigor) [inline Main.hs:21]: It's better not to name variables that you aren't gonna use.

File: Main.hs [hunk 6]
```diff
@@ -7,148 +7,150 @@
...
+import Data.List qualified as L
 import Text.Printf (printf)
 
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex = zipWith (*) [0 ..]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex a = zipWith (^) a [0 ..]
```
Comment (kajigor) [inline Main.hs:17]: You could have avoided introducing the named variable by using `flip`

**Comment by kajigor** [comment]
Good.

**Pull request closed without merging**